TESTS/mbed_hal/us_ticker/main.cpp

```diff
@@ -0,0 +1,114 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "greentea-client/test_env.h"
+
+#include "mbed.h"
+#include "us_ticker_api.h"
+#include "ticker_api.h"
+
+using namespace utest::v1;
+
+namespace {
+    volatile bool complete;
+}
+
+void test_us_ticker_init(void)
+{
+    us_ticker_init();
+    uint32_t previous = us_ticker_read();
+
+    us_ticker_init();
+    uint32_t current = us_ticker_read();
+    TEST_ASSERT_TRUE_MESSAGE(previous <= current, "us_ticker_init() changed the counter");
+
+    previous = us_ticker_read();
+    us_ticker_init();
+    current = us_ticker_read();
+    TEST_ASSERT_TRUE_MESSAGE(previous <= current, "us_ticker_init() changed the counter");
+}
+
+void test_us_ticker_read(void)
+{
+    // this test assumes that to wrap around we would need to run >60 minutes
+    // therefore wrapping should not happen - previous <= current
+    const uint32_t test_loop = 15000;
+    uint32_t previous = us_ticker_read();
+    uint32_t current;
+    for (uint32_t i = 0UL; i < test_loop; i++) {
+        current = us_ticker_read();
+        TEST_ASSERT_TRUE_MESSAGE(previous <= current, "us ticker counter wrapped around");
+    }
+}
+
+void test_us_ticker_read_loop(void)
+{
+    uint32_t future_time = us_ticker_read() + 10000UL;
+    while (us_ticker_read() < future_time);
+    TEST_ASSERT_TRUE_MESSAGE(future_time <= us_ticker_read(), "Future time is in the past");
+}
+
+void us_ticker_handler_test(uint32_t id)
+{
+    complete = true;
+}
+
+void test_us_ticker_set_interrupt(void)
+{
+    // we use here ticker API to set properly callback to set our complete flag
+    const ticker_data_t *us_ticker = get_us_ticker_data();
+    ticker_event_t event;
+
+    complete = false;
+    ticker_set_handler(us_ticker, us_ticker_handler_test);
+    timestamp_t future_timestamp = us_ticker_read() + 5000UL;
+    ticker_insert_event_us(us_ticker, &event, future_timestamp, 1);
```

Comment:
That was the intention but current us ticker API does not provide this, the handler calls directly upper layer, and there is not API to redirect this. Only the upper layer provides that.

We can set interrupt, but handling ISR is in the handler that calls the upper layer. We can remove this test, and the upper layer then exercises set_interrupt() ?